### DISCUSSION TITLE
4.9 on FreeBSD 15.0-CURRENT with ZFS compression: max_size and associated measurements

### DISCUSSION THREAD

**Discussion by grahamperrin** [ccache/ccache#1438] [Support]
I see this commit on the `4.9` branch: 

- https://github.com/ccache/ccache/commit/6fd650cf93c6eb2b96eccd2baa301fb515d7852d

Please: is any of what's below a bug? 

Essentially: 

- max_size = 25.0G
- 64.9 / 25.0 (259.4%) GB reported by the `--show-stats` option of `ccache`
- 44 G measured by `du -hs`. 

Without thinking too deeply, I assume 64.9 _GB_ was compressed to 44 _G_ by ZFS. 

```text
% ccache -s
Cacheable calls:   3269911 / 3591068 (91.06%)
  Hits:             720555 / 3269911 (22.04%)
    Direct:         535539 /  720555 (74.32%)
    Preprocessed:   185016 /  720555 (25.68%)
  Misses:          2549356 / 3269911 (77.96%)
Uncacheable calls:  321111 / 3591068 ( 8.94%)
Errors:                 46 / 3591068 ( 0.00%)
Local storage:
  Cache size (GB):    64.9 /    25.0 (259.4%)
  Cleanups:             19
  Hits:             202708 / 1392739 (14.55%)
  Misses:          1190031 / 1392739 (85.45%)
% cat /internalssd/var/cache/ccache/ccache.conf 
max_size = 25.0G
% echo $CCACHE_DIR
/internalssd/var/cache/ccache
% grep CCACHE_DIR /etc/csh.cshrc 
setenv CCACHE_DIR "/internalssd/var/cache/ccache"
% pkg iinfo ccache4
ccache4-4.9
% date ; time du -hs /internalssd/var/cache/ccache
Tue 23 Apr 2024 02:40:41 BST
 44G    /internalssd/var/cache/ccache
2.942u 97.815s 1:44.97 95.9%    10+167k 100835+0io 0pf+0w
% uname -aKU
FreeBSD mowa219-gjp4-zbook-freebsd 15.0-CURRENT FreeBSD 15.0-CURRENT main-n269594-26f6c148bce2 GENERIC amd64 1500018 1500018
% zfs get compression internalssd
NAME         PROPERTY     VALUE           SOURCE
internalssd  compression  on              default
% grep -i ccache /usr/local/etc/poudriere.conf | grep -v \# | sort
ALLOW_MAKE_JOBS_PACKAGES="ccache pkg py* nvidia-driver-470"
CCACHE_DIR=/internalssd/var/cache/ccache
MAKEWORLDARGS="WITHOUT_LLVM_ASSERTIONS=yes WITH_MALLOC_PRODUCTION=yes -DMALLOC_PRODUCTION WITHOUT_VI=yes WITHOUT_LLVM_TARGET_ALL=yes WITHOUT_TESTS=yes WITH_CCACHE_BUILD=yes -s"
% 
```

### Background 

<https://www.freshports.org/devel/ccache4/> port added 2023-03-13. 

<https://old.reddit.com/r/freebsd/comments/1c98msp/-/l0ou2b2/?context=1> with no change to the size (measured by `du -hs`) following a manual cleanup. 

### See also

<https://www.freshports.org/devel/ccache/> for what's now the inferior version of ccache. 

Reported 2022-02-04 (before version 4 was ported): 

- FreeBSD bug [261705 – devel/ccache when using zfs the cache size reported by ccache is incorrect; with ufs it is correct](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=261705)

**Comment by jrosdahl** [comment]
> I see this commit on the 4.9 branch:
>
> * https://github.com/ccache/ccache/commit/6fd650cf93c6eb2b96eccd2baa301fb515d7852d

OK. Why do you mention this commit?

> Please: is any of what's below a bug?

I don't think it's possible to tell if there is a bug without knowing what events led up to the present state.

> Without thinking too deeply, I assume 64.9 GB was compressed to 44 G by ZFS.

Sounds possible if you have disabled ccache's own compression. What does `du -hs --apparent-size` say?

> 64.9 / 25.0 (259.4%) GB reported by the --show-stats option of ccache

This implies that no compilation (leading to a cache miss) or `ccache -c` call has been performed after the max size has been lowered. Or maybe: An older ccache version has put data there in a way that ccache 4.9 doesn't know about.
